File: src/Sample/Flutter.Net/AppBarTextStylesDemoPage.cs
using System;
using Avalonia;
using Avalonia.Media;
using Flutter.Material;
using Flutter.Rendering;
using Flutter.Widgets;

// Dart parity source (reference): dart_sample/lib/app_bar_text_styles_demo_page.dart (exact sample parity)

namespace Flutter.Net;

public sealed class AppBarTextStylesDemoPage : StatefulWidget
{
    public override State CreateState()
    {
        return new AppBarTextStylesDemoPageState();
    }
}

internal sealed class AppBarTextStylesDemoPageState : State
{
    private static readonly Color AppBarBackground = Color.Parse("#FF1E3A5F");
    private static readonly Color ForegroundWhite = Colors.White;
    private static readonly Color ForegroundMint = Color.Parse("#FFB8FFF1");
    private static readonly TextStyle ThemeTitleStyle = new(
        FontSize: 22,
        Color: Color.Parse("#FF90E0EF"),
        FontWeight: FontWeight.Bold);
    private static readonly TextStyle ThemeToolbarStyle = new(
        FontSize: 14,
        Color: Color.Parse("#FFF4A261"),
        FontWeight: FontWeight.SemiBold);
    private static readonly TextStyle WidgetTitleStyle = new(
        FontSize: 19,
        Color: Color.Parse("#FF8E44AD"),
        FontWeight: FontWeight.Normal);
    private static readonly TextStyle WidgetToolbarStyle = new(
        FontSize: 16,
        Color: Color.Parse("#FFE63946"),
        FontWeight: FontWeight.Bold);

    private Color _foregroundColor = ForegroundWhite;
    private bool _themeTitleStyleEnabled = true;
    private bool _themeToolbarStyleEnabled = true;
    private bool _widgetTitleOverrideEnabled;
    private bool _widgetToolbarOverrideEnabled;

    public override Widget Build(BuildContext context)
    {
        return new Column(
            crossAxisAlignment: CrossAxisAlignment.Stretch,
            spacing: 10,
            children:
            [
                new Text("AppBar title/toolbar text styles", fontSize: 20, color: Colors.Black),
                new Text(
                    "Runtime probe for titleTextStyle and toolbarTextStyle precedence: widget override > appBarTheme > foreground fallback.",
                    fontSize: 14,
                    color: Colors.DimGray),
                new Row(
                    spacing: 8,
                    children:
                    [
                        BuildButton("fg white", () => SetForeground(ForegroundWhite), width: 92, colorHex: "#FFDCE3ED"),
                        BuildButton("fg mint", () => SetForeground(ForegroundMint), width: 92, colorHex: "#FFDCE3ED"),
                    ]),
                new Row(
                    spacing: 8,
                    children:
                    [
                        BuildButton("theme title off", () => SetThemeTitleStyleEnabled(false), width: 114, colorHex: "#FFE9F5EC"),
                        BuildButton("theme title on", () => SetThemeTitleStyleEnabled(true), width: 114, colorHex: "#FFE9F5EC"),
                    ]),
                new Row(
                    spacing: 8,
                    children:
                    [
                        BuildButton("theme toolbar off", () => SetThemeToolbarStyleEnabled(false), width: 126, colorHex: "#FFF3E8D8"),
                        BuildButton("theme toolbar on", () => SetThemeToolbarStyleEnabled(true), width: 126, colorHex: "#FFF3E8D8"),
                    ]),
                new Row(
                    spacing: 8,
                    children:
                    [
                        BuildButton("widget title off", () => SetWidgetTitleOverrideEnabled(false), width: 118, colorHex: "#FFE4ECF7"),
                        BuildButton("widget title on", () => SetWidgetTitleOverrideEnabled(true), width: 118, colorHex: "#FFE4ECF7"),
                    ]),
                new Row(
                    spacing: 8,
                    children:
                    [
                        BuildButton("widget toolbar off", () => SetWidgetToolbarOverrideEnabled(false), width: 130, colorHex: "#FFF5E8ED"),
                        BuildButton("widget toolbar on", () => SetWidgetToolbarOverrideEnabled(true), width: 130, colorHex: "#FFF5E8ED"),
                    ]),
                new Text(
                    $"fg={ColorLabel(_foregroundColor)}, themeTitle={OnOff(_themeTitleStyleEnabled)}, themeToolbar={OnOff(_themeToolbarStyleEnabled)}, widgetTitle={OnOff(_widgetTitleOverrideEnabled)}, widgetToolbar={OnOff(_widgetToolbarOverrideEnabled)}, expectedTitle={DescribeStyle(ResolveExpectedTitleStyle(), _foregroundColor)}, expectedActions={DescribeStyle(ResolveExpectedToolbarStyle(), _foregroundColor)}",
                    fontSize: 12,
                    color: Colors.DarkSlateGray),
                new Container(
                    width: 320,
                    color: Color.Parse("#FFE7EDF6"),
                    padding: new Thickness(10),
                    child: new Column(
                        crossAxisAlignment: CrossAxisAlignment.Stretch,
                        spacing: 8,
                        children:
                        [
                            new Text("Themed app bar preview", fontSize: 11, color: Colors.DimGray),
                            BuildThemedPreview(),
                            new Text("Default app bar reference (no text style overrides)", fontSize: 11, color: Colors.DimGray),
                            BuildDefaultReferencePreview(),
                        ])),
            ]);
    }

    private Widget BuildThemedPreview()
    {
        return new Theme(
            data: ThemeData.Light with
            {
                AppBarTheme = new AppBarThemeData(
                    BackgroundColor: AppBarBackground,
                    ForegroundColor: _foregroundColor,
                    TitleTextStyle: _themeTitleStyleEnabled ? ThemeTitleStyle : null,
                    ToolbarTextStyle: _themeToolbarStyleEnabled ? ThemeToolbarStyle : null),
            },
            child: new AppBar(
                titleText: "Theme text styles",
                titleTextStyle: _widgetTitleOverrideEnabled ? WidgetTitleStyle : null,
                toolbarTextStyle: _widgetToolbarOverrideEnabled ? WidgetToolbarStyle : null,
                actions:
                [
                    new Text("A1"),
                    new Text("A2"),
                ]));
    }

    private Widget BuildDefaultReferencePreview()
    {
        return new AppBar(
            titleText: "Default text styles",
            backgroundColor: AppBarBackground,
            foregroundColor: _foregroundColor,
            actions:
            [
                new Text("A1"),
                new Text("A2"),
            ]);
    }

    private Widget BuildButton(string label, Action onTap, double width, string colorHex)
    {
        return new SizedBox(
            width: width,
            child: new CounterTapButton(
                label: label,
                onTap: onTap,
                background: Color.Parse(colorHex),
                foreground: Colors.Black,
                fontSize: 12,
                padding: new Thickness(10, 8)));
    }

    private void SetForeground(Color value)
    {
        SetState(() => _foregroundColor = value);
    }

    private void SetThemeTitleStyleEnabled(bool value)
    {
        SetState(() => _themeTitleStyleEnabled = value);
    }

    private void SetThemeToolbarStyleEnabled(bool value)
    {
        SetState(() => _themeToolbarStyleEnabled = value);
    }

    private void SetWidgetTitleOverrideEnabled(bool value)
    {
        SetState(() => _widgetTitleOverrideEnabled = value);
    }

    private void SetWidgetToolbarOverrideEnabled(bool value)
    {
        SetState(() => _widgetToolbarOverrideEnabled = value);
    }

    private TextStyle? ResolveExpectedTitleStyle()
    {
        if (_widgetTitleOverrideEnabled)
        {
            return WidgetTitleStyle;
        }

        if (_themeTitleStyleEnabled)
        {
            return ThemeTitleStyle;
        }

        return null;
    }

    private TextStyle? ResolveExpectedToolbarStyle()
    {
        if (_widgetToolbarOverrideEnabled)
        {
            return WidgetToolbarStyle;
        }

        if (_themeToolbarStyleEnabled)
        {
            return ThemeToolbarStyle;
        }

        return null;
    }

    private static string DescribeStyle(TextStyle? style, Color fallbackColor)
    {
        var color = style?.Color ?? fallbackColor;
        var fontSize = style?.FontSize;
        var sizeLabel = fontSize.HasValue ? fontSize.Value.ToString("0") : "base";
        return $"{ColorLabel(color)}:{sizeLabel}";
    }

    private static string OnOff(bool value) => value ? "on" : "off";

    private static string ColorLabel(Color color) => $"#{color.A:X2}{color.R:X2}{color.G:X2}{color.B:X2}";
}

File: dart_sample/lib/app_bar_text_styles_demo_page.dart
import 'package:flutter/material.dart';

import 'counter_widgets.dart';

class AppBarTextStylesDemoPage extends StatefulWidget {
  const AppBarTextStylesDemoPage({super.key});

  @override
  State<AppBarTextStylesDemoPage> createState() =>
      _AppBarTextStylesDemoPageState();
}

class _AppBarTextStylesDemoPageState extends State<AppBarTextStylesDemoPage> {
  static const Color _appBarBackground = Color(0xFF1E3A5F);
  static const Color _foregroundWhite = Colors.white;
  static const Color _foregroundMint = Color(0xFFB8FFF1);
  static const TextStyle _themeTitleStyle = TextStyle(
    fontSize: 22,
    color: Color(0xFF90E0EF),
    fontWeight: FontWeight.bold,
  );
  static const TextStyle _themeToolbarStyle = TextStyle(
    fontSize: 14,
    color: Color(0xFFF4A261),
    fontWeight: FontWeight.w600,
  );
  static const TextStyle _widgetTitleStyle = TextStyle(
    fontSize: 19,
    color: Color(0xFF8E44AD),
    fontWeight: FontWeight.normal,
  );
  static const TextStyle _widgetToolbarStyle = TextStyle(
    fontSize: 16,
    color: Color(0xFFE63946),
    fontWeight: FontWeight.bold,
  );

  Color _foregroundColor = _foregroundWhite;
  bool _themeTitleStyleEnabled = true;
  bool _themeToolbarStyleEnabled = true;
  bool _widgetTitleOverrideEnabled = false;
  bool _widgetToolbarOverrideEnabled = false;

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      spacing: 10,
      children: <Widget>[
        const Text(
          'AppBar title/toolbar text styles',
          style: TextStyle(fontSize: 20, color: Colors.black),
        ),
        const Text(
          'Runtime probe for titleTextStyle and toolbarTextStyle precedence: widget override > appBarTheme > foreground fallback.',
          style: TextStyle(fontSize: 14, color: Colors.black54),
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            _buildButton(
              label: 'fg white',
              onTap: () => _setForeground(_foregroundWhite),
              width: 92,
              background: const Color(0xFFDCE3ED),
            ),
            _buildButton(
              label: 'fg mint',
              onTap: () => _setForeground(_foregroundMint),
              width: 92,
              background: const Color(0xFFDCE3ED),
            ),
          ],
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            _buildButton(
              label: 'theme title off',
              onTap: () => _setThemeTitleStyleEnabled(false),
              width: 114,
              background: const Color(0xFFE9F5EC),
            ),
            _buildButton(
              label: 'theme title on',
              onTap: () => _setThemeTitleStyleEnabled(true),
              width: 114,
              background: const Color(0xFFE9F5EC),
            ),
          ],
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            _buildButton(
              label: 'theme toolbar off',
              onTap: () => _setThemeToolbarStyleEnabled(false),
              width: 126,
              background: const Color(0xFFF3E8D8),
            ),
            _buildButton(
              label: 'theme toolbar on',
              onTap: () => _setThemeToolbarStyleEnabled(true),
              width: 126,
              background: const Color(0xFFF3E8D8),
            ),
          ],
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            _buildButton(
              label: 'widget title off',
              onTap: () => _setWidgetTitleOverrideEnabled(false),
              width: 118,
              background: const Color(0xFFE4ECF7),
            ),
            _buildButton(
              label: 'widget title on',
              onTap: () => _setWidgetTitleOverrideEnabled(true),
              width: 118,
              background: const Color(0xFFE4ECF7),
            ),
          ],
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            _buildButton(
              label: 'widget toolbar off',
              onTap: () => _setWidgetToolbarOverrideEnabled(false),
              width: 130,
              background: const Color(0xFFF5E8ED),
            ),
            _buildButton(
              label: 'widget toolbar on',
              onTap: () => _setWidgetToolbarOverrideEnabled(true),
              width: 130,
              background: const Color(0xFFF5E8ED),
            ),
          ],
        ),
        Text(
          'fg=${_colorLabel(_foregroundColor)}, themeTitle=${_onOff(_themeTitleStyleEnabled)}, themeToolbar=${_onOff(_themeToolbarStyleEnabled)}, widgetTitle=${_onOff(_widgetTitleOverrideEnabled)}, widgetToolbar=${_onOff(_widgetToolbarOverrideEnabled)}, expectedTitle=${_describeStyle(_resolveExpectedTitleStyle(), _foregroundColor)}, expectedActions=${_describeStyle(_resolveExpectedToolbarStyle(), _foregroundColor)}',
          style: const TextStyle(fontSize: 12, color: Colors.blueGrey),
        ),
        Container(
          width: 320,
          color: const Color(0xFFE7EDF6),
          padding: const EdgeInsets.all(10),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.stretch,
            spacing: 8,
            children: <Widget>[
              const Text(
                'Themed app bar preview',
                style: TextStyle(fontSize: 11, color: Colors.black54),
              ),
              _buildThemedPreview(),
              const Text(
                'Default app bar reference (no text style overrides)',
                style: TextStyle(fontSize: 11, color: Colors.black54),
              ),
              _buildDefaultReferencePreview(),
            ],
          ),
        ),
      ],
    );
  }

  Widget _buildThemedPreview() {
    return Theme(
      data: ThemeData.light().copyWith(
        appBarTheme: AppBarTheme(
          backgroundColor: _appBarBackground,
          foregroundColor: _foregroundColor,
          titleTextStyle: _themeTitleStyleEnabled ? _themeTitleStyle : null,
          toolbarTextStyle: _themeToolbarStyleEnabled
              ? _themeToolbarStyle
              : null,
        ),
      ),
      child: AppBar(
        title: const Text('Theme text styles'),
        titleTextStyle: _widgetTitleOverrideEnabled ? _widgetTitleStyle : null,
        toolbarTextStyle: _widgetToolbarOverrideEnabled
            ? _widgetToolbarStyle
            : null,
        actions: const <Widget>[Text('A1'), Text('A2')],
      ),
    );
  }

  Widget _buildDefaultReferencePreview() {
    return AppBar(
      title: const Text('Default text styles'),
      backgroundColor: _appBarBackground,
      foregroundColor: _foregroundColor,
      actions: const <Widget>[Text('A1'), Text('A2')],
    );
  }

  Widget _buildButton({
    required String label,
    required VoidCallback onTap,
    required double width,
    required Color background,
  }) {
    return SizedBox(
      width: width,
      child: CounterTapButton(
        label: label,
        onTap: onTap,
        background: background,
        foreground: Colors.black,
        fontSize: 12,
        padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 8),
      ),
    );
  }

  void _setForeground(Color value) {
    setState(() {
      _foregroundColor = value;
    });
  }

  void _setThemeTitleStyleEnabled(bool value) {
    setState(() {
      _themeTitleStyleEnabled = value;
    });
  }

  void _setThemeToolbarStyleEnabled(bool value) {
    setState(() {
      _themeToolbarStyleEnabled = value;
    });
  }

  void _setWidgetTitleOverrideEnabled(bool value) {
    setState(() {
      _widgetTitleOverrideEnabled = value;
    });
  }

  void _setWidgetToolbarOverrideEnabled(bool value) {
    setState(() {
      _widgetToolbarOverrideEnabled = value;
    });
  }

  TextStyle? _resolveExpectedTitleStyle() {
    if (_widgetTitleOverrideEnabled) {
      return _widgetTitleStyle;
    }

    if (_themeTitleStyleEnabled) {
      return _themeTitleStyle;
    }

    return null;
  }

  TextStyle? _resolveExpectedToolbarStyle() {
    if (_widgetToolbarOverrideEnabled) {
      return _widgetToolbarStyle;
    }

    if (_themeToolbarStyleEnabled) {
      return _themeToolbarStyle;
    }

    return null;
  }

  static String _describeStyle(TextStyle? style, Color fallbackColor) {
    final Color color = style?.color ?? fallbackColor;
    final String sizeLabel = style?.fontSize?.toStringAsFixed(0) ?? 'base';
    return '${_colorLabel(color)}:$sizeLabel';
  }

  static String _onOff(bool value) => value ? 'on' : 'off';

  static String _colorLabel(Color color) {
    final int value = color.toARGB32();
    return '#${value.toRadixString(16).toUpperCase().padLeft(8, '0')}';
  }
}
